src/nodes/mod.rs

```diff
@@ -74,9 +75,71 @@ impl From<nix::Error> for KernelError {
     }
 }
 
+/// Container for new attribute values to set on any kind of node.
+pub struct AttrDelta {
+    pub mode: Option<sys::stat::Mode>,
+    pub uid: Option<unistd::Uid>,
+    pub gid: Option<unistd::Gid>,
+    pub atime: Option<sys::time::TimeVal>,
+    pub mtime: Option<sys::time::TimeVal>,
+    pub size: Option<u64>,
+}
+
 /// Generic result type for of all node operations.
 pub type NodeResult<T> = Result<T, KernelError>;
 
+/// Updates the metadata of a file given a delta of attributes.
+///
+/// This is a helper function to implement `Node::setattr` for the various node types.  The caller
+/// is responsible for obtaining the updated metadata of the underlying file to return it to the
+/// kernel, as computing the updated metadata here would be inaccurate.  The reason is that we don't
+/// track ctimes ourselves so any modifications to the file cause the backing ctime to be updated
+/// and we need to obtain it.
+/// TODO(jmmv): Compute the modified fuse::FileAttr and return it once we track ctimes.
+///
+/// This tries to apply as many properties as possible in case of errors.  When errors occur,
+/// returns the first that was encountered.
+pub fn setattr(path: &Path, attr: &fuse::FileAttr, delta: &AttrDelta) -> Result<(), nix::Error> {
+    let mut result = Ok(());
+
+    if let Some(mode) = delta.mode {
+        result = result.and(
+            sys::stat::fchmodat(None, path, mode, sys::stat::FchmodatFlags::NoFollowSymlink));
+    }
+
+    if delta.uid.is_some() || delta.gid.is_some() {
+        result = result.and(
+            unistd::fchownat(None, path, delta.uid, delta.gid,
```

Comment:
I don't suggest supporting chown/chgrp.

Reply:
Heh, interesting thought. What's your rationale though?

(I'm going to leave this in for now given that the Go version supports it and there are integration tests covering this case. But I could add a TODO.)

Reply:
It's just an invitation to security bugs. Non-privileged workloads can almost never use chown (and rarely use chgrp) legitimately anyway.

Reply:
Now that I think of it, the reason this can't be removed is because I'm planning on (also) using sandboxfs to contain the build of pkgsrc packages (via https://github.com/jmmv/pkg_comp, which currently uses bindfs). Building arbitrary software requires root privileges, and the builds actually end up calling into chown/chgrp...

Reply:
Debian uses fakeroot to avoid needing privileges...